# Genome and Transcriptome Analysis

This Northeastern academic project contains the code and documentation for a comprehensive genome and transcriptome analysis. Below are the key tasks and tools used in this analysis, along with links to the corresponding GitHub repositories.
  
Skills:**Shell, Python, High-Performance Computing Platform**  

## Genome Assembly and Quality Evaluation

Performed genome assembly and quality evaluation using Trimmomatic, Spades, and Quast.  
[Genome Assembly](https://github.com/penghy27/NU-NGS_analysis/tree/main/GenomeAssembly)
  
## RNA-seq Analysis

Conducted RNA-seq read alignment (GMAP, GSNAP), reference-guided and de novo assembly (Trinity) for *Aiptasia pallida*.  
[Short Read Alignment](https://github.com/penghy27/NU-NGS_analysis/tree/main/RNAseqAnalysis)  
[Transcriptome Assembly](https://github.com/penghy27/NGS_analysis/tree/main/TranscriptomeAssembly)

## Protein Sequence Prediction and Annotation

Predicted protein sequences and candidate coding regions with BLAST and TransDecoder.
Aligned protein sequences with SwissPro.  
[Protein Sequence Analysis](https://github.com/penghy27/NU-NGS_analysis/tree/main/ProteinSequenceAnalysis)

## Functional Annotation

Utilized KEGG databases for comprehensive annotation of protein sequences.  
[Functional Annotation](https://github.com/penghy27/NU-NGS_analysis/tree/main/FunctionalAnnotation)

  
Please refer to the individual repositories linked above for detailed code, data, and documentation related to each specific analysis task.  

Other bioinformatics projects can be found [here](https://github.com/penghy27/NU_academic_projects.git).
